### PR TITLE
[JS] feat: Azure AI Search data source

### PR DESCRIPTION
**Pull request by singhk97**
## Linked issues

closes: #1360 #1267 #1266 (issue number)

## Details
Introducing a sample to showcase how to integrate the Azure AI Search data source.

#### Change details
- Included scripts to create/delete index
- Included Teams Test tool
- Azure AI Search data source

## Attestation Checklist

- [x] My code follows the style guidelines of this project

- I have checked for/fixed spelling, linting, and other errors
- I have commented my code for clarity
- I have made corresponding changes to the documentation (updating the doc strings in the code is sufficient)
- My changes generate no new warnings
- I have added tests that validates my changes, and provides sufficient test coverage. I have tested with:
  - Local testing
  - E2E testing in Teams
- New and existing unit tests pass locally with my changes

### CODE REVIEW

File: js/samples/08.datasource.azureAISearch/appPackage/manifest.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (corinagum) [inline js/samples/08.datasource.azureAISearch/appPackage/manifest.json:1]: Above: use different icon for sample manifest

Comment (singhk97) [reply js/samples/08.datasource.azureAISearch/appPackage/manifest.json:1]: updated

File: js/samples/08.datasource.azureAISearch/infra/azure.parameters.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        "value": "${{BOT_ID}}"
+      },
+      "botAadAppClientSecret": {
+        "value": "${{SECRET_BOT_PASSWORD}}"
+      },
+      "webAppSKU": {
+        "value": "B1"
+      },
+      "botDisplayName": {
+        "value": "TeamsChef"
```
Comment (corinagum) [inline js/samples/08.datasource.azureAISearch/infra/azure.parameters.json:18]: Fix name

File: js/samples/08.datasource.azureAISearch/teamsapp.local.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      name: TeamsChef
+    writeToEnvironmentFile:
+      botId: BOT_ID
+      botPassword: SECRET_BOT_PASSWORD
+
+  - uses: botFramework/create # Create or update the bot registration on dev.botframework.com
+    with:
+      botId: ${{BOT_ID}}
+      name: TeamsChef
+      messagingEndpoint: ${{BOT_ENDPOINT}}/api/messages
```
Comment (corinagum) [inline js/samples/08.datasource.azureAISearch/teamsapp.local.yml:24]: Fix name

File: js/samples/08.datasource.azureAISearch/teamsapp.yml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    with:
+      name: TeamsChef-${{TEAMSFX_ENV}} # Teams app name
+    writeToEnvironmentFile:
+      # Write the information of installed dependencies into environment file for the specified environment variable(s).
+      teamsAppId: TEAMS_APP_ID
+
+  - uses: botAadApp/create # Creates a new AAD app for Bot Registration.
+    with:
+      name: TeamsChef
+    writeToEnvironmentFile:
```
Comment (corinagum) [inline js/samples/08.datasource.azureAISearch/teamsapp.yml:20]: Rename

**Comment by lilyydu** [comment]
Not sure if this is in the plan already but do we want to add this into the CONCEPTS docs to break down the components + outline steps on how to customize it (beyond restaurant data)? 

I also managed to test the samples in Python without TestTool if we prefer excluding it!

**Comment by singhk97** [comment]
> Not sure if this is in the plan already but do we want to add this into the CONCEPTS docs to break down the components + outline steps on how to customize it (beyond restaurant data)?
> 
> I also managed to test the samples in Python without TestTool if we prefer excluding it!

Good point Lily. I'll open up a PR for this once all the implementations are checked in. Thanks for bringing this up.

File: js/samples/08.datasource.azureAISearch/package.json [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        "dev:teamsfx:testtool": "env-cmd --silent -f .env npm run dev",
+        "dev:teamsfx:launch-testtool": "env-cmd --silent -f env/.env.testtool teamsapptester start",
+        "dev": "nodemon --watch ./src --exec node --inspect=9239 --signal SIGINT -r ts-node/register ./src/index.ts",
+        "build": "tsc --build",
+        "clean": "rimraf node_modules lib",
+        "lint": "eslint **/src/**/*.{j,t}s{,x} --fix --no-error-on-unmatched-pattern",
+        "start": "tsc --build && node ./lib/index.js",
+        "test": "echo \"Error: no test specified\" && exit 1",
+        "watch": "nodemon --watch ./src -e ts --exec \"yarn start\"",
+        "indexer:create": "yarn build && env-cmd --silent -f .env node ./lib/indexer/setup.js",
```
Comment (frankqianms) [inline js/samples/08.datasource.azureAISearch/package.json:22]: should be:
`"indexer:create": "yarn build && env-cmd --silent -f .env node ./lib/indexers/setup.js"`

File: js/samples/08.datasource.azureAISearch/package.json [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        "indexer:delete": "yarn build && env-cmd --silent -f .env node ./lib/indexer/delete.js"
```
Comment (frankqianms) [inline js/samples/08.datasource.azureAISearch/package.json:23]: same here

File: js/samples/08.datasource.azureAISearch/src/AzureAISearchDataSource.ts [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+     */
+    name: string;
+
+    /**
+     * Name of the Azure AI Search index.
+     */
+    indexName: string;
+
+    /**
+     * Azure OpenAI API key. AzureO
```
Comment (lilyydu) [inline js/samples/08.datasource.azureAISearch/src/AzureAISearchDataSource.ts:41]: Extra "AzureO"

**Review by lilyydu — APPROVED** [review]
small typo but otherwise lgtm :)

File: js/samples/08.datasource.azureAISearch/src/AzureAISearchDataSource.ts [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+        const selectedFields = [
+            'restaurantId',
+            'restaurantName',
+            'description',
+            'category',
+            'tags',
+            'rating',
+            'location',
+            'link',
```
Comment (julbrs) [inline js/samples/08.datasource.azureAISearch/src/AzureAISearchDataSource.ts:132]: `link` is not defined in the index, so it fails actually with this. Suggesting to remove this field from the selection ;)

BTW thanks for this big PR, I was about to start working on that, it saves me a lot of time,

**Review by lilyydu — APPROVED** [review]

**Pull request merged**